### PR TITLE
Add parquet compressor using columnify

### PR DESCRIPTION
**Pull request by okkez**
How to use:

Install [columnify](https://github.com/reproio/columnify).

The sample configuration for parquet compressor.

```
<match>
  @id s3-parquet
  @type s3

  s3_region ap-northeast-1
  s3_bucket xxx
  store_as parquet
  <compress>
    parquet_compression_codec snappy
    record_type msgpack
    schema_type avro
    schema_file /path/to/log.avsc
  </compress>

  <format>
    @type msgpack
  </format>
</match>
```

log.avsc is like following:

```json
{
  "name": "AccessLog",
  "type": "record",
  "fields": [
    { "name": "container_id", "type": "string" },
    { "name": "method", "type": "string" },...
  ]
}
```

See https://avro.apache.org/docs/current/spec.html for more details about avro schema.

Notice: 
columnify's memory usage is proportional to file size. For example, columnify consumes about 750MB memory (RSS) while processing 128MB msgpack.

See also #221

### CODE REVIEW

**Comment by repeatedly** [comment]
Looks good to me.
Need feedback from other users.

**Comment by scrwr** [comment]
I'd actually really love to test it. But. We use bundler in our build process for installing all plugins and fluentd unfortunately stopped importing bundler installed git sources, so I end up with "Unknown output plugin 's3'" if I add this to our Gemfile

`gem 'fluent-plugin-s3', git: 'https://github.com/okkez/fluent-plugin-s3.git', branch: 'add-parquet-compressor'`

Can you help with getting this to work without completely workarounding the bundle install.

**Comment by freelancerlucas** [comment]
thank you for this, I really appreciate that

File: lib/fluent/plugin/s3_compressor_parquet.rb
Comment (davemt) [inline lib/fluent/plugin/s3_compressor_parquet.rb:60]: For the majority of cases this error should not be retried. For example, the schema may not match some new log format coming through.

Given that it should probably raise `Fluent::UnrecoverableError` here. See https://docs.fluentd.org/buffer#handling-unrecoverable-errors

**Review by davemt — CHANGES_REQUESTED** [review]

File: lib/fluent/plugin/s3_compressor_parquet.rb
Comment (okkez) [reply lib/fluent/plugin/s3_compressor_parquet.rb]: Sorry for the late response. I 've pushed changes about this.

**Comment by okkez** [comment]
This parquet compressor has worked fine for the recent 9 months on our system.

Can I merge this PR?

**Comment by okkez** [comment]
@repeatedly @ganmacs @kenhys @ashie Can you merge this PR and release the new version?

**Comment by ashie** [comment]
I'll merge it after waiting for comments from other maintainers for a few days.

File: README.md
Comment (ashie) [inline README.md:413]: @okkez 
It seems to require more explanation in README.md to use this compressor compared with the other compressors. Otherwise users need to read the source code or find this issue. Could you add it?

**Review by ashie — CHANGES_REQUESTED** [review]
> The sample configuration for parquet compressor.
> 
> ```
> <match>
>   @id s3-parquet
>   @type s3
> 
>   s3_region ap-northeast-1
>   s3_bucket xxx
>   <compress>
>     parquet_compression_codec snappy
>     record_type msgpack
>     schema_type avro
>     schema_file /path/to/log.avsc
>   </compress>
> 
>   <format>
>     @type msgpack
>   </format>
> </match>
> ```

@okkez `store_as parquet` is also required, isnt it?

**Comment by okkez** [comment]
> @okkez `store_as parquet` is also required, isnt it?

Yes. It is required.

File: README.md
Comment (okkez) [reply README.md:413]: OK, I'll add some more explanation.

Comment (ashie) [reply README.md:413]: Thanks!
Hmm, it's not your fault but the structure of the sentence is a bit hard to understand.
I'll try to refine it in another PR.

File: lib/fluent/plugin/s3_compressor_parquet.rb
Comment (cosmo0920) [inline lib/fluent/plugin/s3_compressor_parquet.rb:27]: Nits: Braces is required for readability?

```ruby
if [:lzo, :brotli, :lz4].include?(@compress.parquet_compression_codec)
```

File: README.md
Comment (cosmo0920) [inline README.md:429]: lzo, brotli, lz4 are still unsupported in parquet compressor:
https://github.com/fluent/fluent-plugin-s3/pull/338/files#diff-a86f63eacf5404bd0230b96dcc02a8ac85d0f45d1c36cd6a1f5b5ad25444a1ddR27-R29
Should we add note for this limitations?

File: lib/fluent/plugin/s3_compressor_parquet.rb
Comment (okkez) [reply lib/fluent/plugin/s3_compressor_parquet.rb:27]: Ah, I forgot it.

**Comment by ashie** [comment]
We'll refine the document at #373

**Comment by ashie** [comment]
Thank you for your work!

**Comment by sunnymeska-w** [comment]
the compressor dosnt support list type in schema file. please add support for input type list

**Pull request merged**
